<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Deconvolution</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Deconvolution</h1>



<div id="deconvolution-using-rctd" class="section level1">
<h1>Deconvolution using RCTD</h1>
<p>Here we show how to perform cell type deconvolution using RCTD
(Robust Cell Type Decomposition).<br />
The first step is to read in the reference dataset and create a
reference object</p>
<details>
<summary>
Step 1: Reading in a Reference Dataset
</summary>
<pre><code>###Step 0: Packages
library(spacexr)
library(Matrix)
library(Seurat) 
##download spacexr is not installed
#library(devtools)
#devtools::install_github(&quot;dmcable/spacexr&quot;, build_vignettes = FALSE)
###STEP 1: Read in reference dataset
#Read in reference
data(&quot;liver_met_ref&quot;)
refr = liver_met_ref
#get cell types of reference dataset
cell_types = Idents(refr)
#drop levels 
cell_types = droplevels(cell_types)
#get raw data matrix
count_raw &lt;- refr@assays$RNA@counts
# make reference dataset
reference &lt;- Reference(count_raw, cell_types = cell_types)</code></pre>
</details>
The second step is to read in the spatial data.
<details>
<summary>
Step 2: Reading in Spatial Data
</summary>
<pre><code>###STEP 2: Read in spatial data
#Read in seurat data. In practice, use seurat function to read in data
data(&quot;liver_met_seurat_object&quot;)
seurat_object = liver_met_seurat_object
#get counts matrix 
counts = seurat_object@assays$Spatial@counts
#save gene and cell names for later
genes = colnames(counts)
spots = rownames(counts)
#reformat counts matrix to sparse matrix 
counts = as(counts,&#39;sparseMatrix&#39;)
#name column and row names
colnames(counts) = genes
rownames(counts) = spots
#get coordinate matrix
coord = GetTissueCoordinates(seurat_object)
#make spatial puck
puck &lt;- SpatialRNA(coord, counts)</code></pre>
</details>
<p>The third step is to create an RCTD object and perform RCTD. Here we
set the max number of cell types in a spot to be 4. Please refer to the
documentation for information on other parameters</p>
<details>
<summary>
Step 3: Creating an RCTD Object
</summary>
<pre><code>###STEP 3: create an RCTD object
#create an RCTD object. Here we set the max number of cell types in a spot to be 4.
#see documentation for other parameter choices
myRCTD &lt;- create.RCTD(puck, reference, max_cores = 1, UMI_min = 0,MAX_MULTI_TYPES = 4)
#Run RCTD
myRCTD &lt;- run.RCTD(myRCTD, doublet_mode = &quot;multi&quot;)</code></pre>
</details>
<p>The fourth step is to reformat your RCTD output into a matrix. This
matrix will be of dimension #spots by #cell types. Each row will contain
the deconvolution estimate for the corresponding spot.</p>
<details>
<summary>
Step 4: Making a Deconvolution Matrix
</summary>
<pre><code>###Step 4: Reformat results into a matrix 
#get unique cell types
CT = unique(cell_types)
#initialize the deconvolution matrix 
deconv_est = matrix(0,nrow(coord),length(CT))
#Column names will be cell types
colnames(deconv_est) = CT
#rownames will be spot names
rownames(deconv_est) = rownames(coord)
#iterate over deconvolution results 
for(j in c(1:length(myRCTD@results))){
  #match cell types found to index of unique cell type vector
  fills = match(myRCTD@results[[j]]$cell_type_list,CT)
  #fill in matrix 
  deconv_est[j,fills] = myRCTD@results[[j]]$sub_weights
  #normalize so that rows sum to 1
  deconv_est[j,] = deconv_est[j,]/sum(deconv_est[j,])
}
#final output
deconv_est</code></pre>
</details>
<p>For more information, visit the <a href="https://github.com/dmcable/spacexr">spacexr</a> github page.</p>
</div>
<div id="single-cell-data" class="section level1">
<h1>Single Cell Data</h1>
<p>If your data is of the single cell resolution, then the deconvolution
matrix is a dummy variable matrix. See <a href="https://www.geeksforgeeks.org/dummy-variables-in-r-programming/">this
link</a> for a tutorial on how to make dummy variable matrices. Note
that your deconvolution matrix and your average expression profile
matrix must have the cell types in the same order.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
