<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Object Creation From A Seurat Object</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Object Creation From A Seurat Object</h1>



Starting with a spatial seurat object, we can make a niche-DE object
with the function ‘CreateNicheDEObjectFromSeurat’. This function takes
in 5 arguments
<details>
<summary>
Arguments
</summary>
<ul>
<li>seurat object: A seurat object
<ul>
<li>assay: The assay from which to retrieve the counts matrix. Will be
taken from slot ‘counts’</li>
<li>library mat: The average expression profile matrix calculated from a
reference dataset</li>
<li>deconv mat: The deconvolution matrix for the spatial dataset</li>
<li>sigma: A list of kernel bandwidths to use for effective niche
calculation. They should be chosen with the knowledge that the
coordinates are scaled such that the median nearest neighbor distance is
100.</li>
</ul>
</details></li>
</ul>
<pre><code>#read in seurat object
data(&#39;liver_met_seurat_object&#39;)
#download average expression profile matrix 
data(&quot;vignette_library_matrix&quot;)
data(&quot;vignette_deconv_mat&quot;)
#make niche-DE object
NDE_obj = CreateNicheDEObjectFromSeurat(liver_met_seurat_object,&#39;Spatial&#39;,
vignette_library_matrix,vignette_deconv_mat,sigma = c(1,100,250))
</code></pre>
<details>
<summary>
What’s in the object?
</summary>
<p>We see that their are 14 slots, 10 of which are populated when making
the nicheDE object. Here we will explain what each slot should contain,
except for the ones prefixed by niche_DE.</p>
<ul>
<li>counts: The RNA count data of the spatial transcriptomics dataset.
The dimension will be #cells/spots by #genes.Genes are filtered out if
they do not exist within the scrna-seq reference dataset.</li>
<li>coord: The spatial coordinates matrix of the spatial transcriptomics
dataset.It is scaled such that the median nearest neighbor distance is
100.</li>
<li>sigma: The kernel bandwidth(s) chosen for calculating the effective
niche. Recommended values will be discussed shortly.</li>
<li>num_cells: A #cells/spots by #cell types matrix indicating the
estimated number of cells of each cell type in each spot.</li>
<li>effective_niche: A list whose length is equal to the length of
sigma. Each element of the list is a matrix of dimension #cells/spots by
#cell types that measures how many of each cell type is in a given
cell/spot’s neighborhood. For more information, please read the
manuscript.</li>
<li>ref_expr: The average expression profile matrix. The dimension is
#cell types by #genes. Each row gives the average expression of each
gene for a given cell type based on the reference dataset supplied.</li>
<li>cell_names: The name of each cell. This will be used if the use
wants to filter cells via the function ‘Filter’</li>
<li>gene_names: The gene names.</li>
<li>batch_ID: The batch ID for each cell/spot. This will be used when
merging objects.</li>
<li>spot_distance: The mean distance between a cell/spot and its nearest
neighbor.This value can be used to inform the choice of sigma.</li>
</ul>
</details>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
