<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Object Creation From A Seurat Object • nicheDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Object Creation From A Seurat Object">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="inverse" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nicheDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/Tutorial.html">Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Kmason23/NicheDE"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Object Creation From A Seurat Object</h1>
            
      

      <div class="d-none name"><code>Object_Creation_Seurat.Rmd</code></div>
    </div>

    
    
Starting with a spatial seurat object, we can make a niche-DE object
with the function ‘CreateNicheDEObjectFromSeurat’. This function takes
in 6 arguments
<details><summary>
Arguments
</summary><ul><li>seurat object: A seurat object
<ul>
<li>assay: The assay from which to retrieve the counts matrix. Will be
taken from slot ‘counts’</li>
<li>library mat: The average expression profile matrix calculated from a
reference dataset</li>
<li>deconv mat: The deconvolution matrix for the spatial dataset</li>
<li>sigma: A list of kernel bandwidths to use for effective niche
calculation. They should be chosen with the knowledge that the
coordinates are scaled such that the median nearest neighbor distance is
100.</li>
<li>Int: Logical if counts data supplied is integer. When performing
niche-DE, Negative binomial regression is performed if True. Linear
regression with a gene specific variance is performed if False. Default
is true.</li>
</ul>
</li></ul></details><pre><code><span><span class="co">#read in seurat object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'liver_met_seurat_object'</span><span class="op">)</span></span>
<span><span class="co">#download average expression profile matrix </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"vignette_library_matrix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"vignette_deconv_mat"</span><span class="op">)</span></span>
<span><span class="co">#make niche-DE object</span></span>
<span><span class="va">NDE_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/CreateNicheDEObjectFromSeurat.html">CreateNicheDEObjectFromSeurat</a></span><span class="op">(</span><span class="va">liver_met_seurat_object</span>,<span class="st">'Spatial'</span>,</span>
<span><span class="va">vignette_library_matrix</span>,<span class="va">vignette_deconv_mat</span>,sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">100</span>,<span class="fl">250</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<details><summary>
What’s in the object?
</summary><p>We see that their are 14 slots, 10 of which are populated when making
the nicheDE object. Here we will explain what each slot should contain,
except for the ones prefixed by niche_DE.</p>
<ul>
<li>counts: The RNA count data of the spatial transcriptomics dataset.
The dimension will be #cells/spots by #genes.Genes are filtered out if
they do not exist within the scrna-seq reference dataset.</li>
<li>coord: The spatial coordinates matrix of the spatial transcriptomics
dataset.It is scaled such that the median nearest neighbor distance is
100.</li>
<li>sigma: The kernel bandwidth(s) chosen for calculating the effective
niche. Recommended values will be discussed shortly.</li>
<li>num_cells: A #cells/spots by #cell types matrix indicating the
estimated number of cells of each cell type in each spot.</li>
<li>effective_niche: A list whose length is equal to the length of
sigma. Each element of the list is a matrix of dimension #cells/spots by
#cell types that measures how many of each cell type is in a given
cell/spot’s neighborhood. For more information, please read the
manuscript.</li>
<li>ref_expr: The average expression profile matrix. The dimension is
#cell types by #genes. Each row gives the average expression of each
gene for a given cell type based on the reference dataset supplied.</li>
<li>cell_names: The name of each cell. This will be used if the use
wants to filter cells via the function ‘Filter’</li>
<li>gene_names: The gene names.</li>
<li>batch_ID: The batch ID for each cell/spot. This will be used when
merging objects.</li>
<li>spot_distance: 100. Coordinates are scaled so that the distance
between a spot and its nearest neighbor is 100.</li>
<li>scale: The scaling to convert our coordinates back to their original
scale</li>
<li>Int: Is the data count data(True) or continuous (False)?</li>
</ul></details></main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
