<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Ligand Receptor Analysis • nicheDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Ligand Receptor Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="inverse" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nicheDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/Tutorial.html">Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Kmason23/NicheDE"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Ligand Receptor Analysis</h1>
            
      

      <div class="d-none name"><code>Ligand_Receptor_Analysis.Rmd</code></div>
    </div>

    
    
<p>To see what ligand-receptor channels are driving niche signals in our
data, we developed niche-LR. Niche-LR uses the <a href="https://www.nature.com/articles/s41592-019-0667-5" class="external-link">niche-net</a>
ligand-target potential matrix to find the top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
downstream targets for each ligand. The niche-DE T-statistics of these
downstream genes are used to calculate ligand activity scores. The top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>M</mi><annotation encoding="application/x-tex">M</annotation></semantics></math>
ligands by ligand activity score and their corresponding receptors are
then tested to confirm expression in the tissue. Niche-LR can be
performed on spot resolution data using the function ‘niche_LR_spot’ and
on single cell resolution data using the function ‘niche_LR_cell’.
‘niche_LR_spot’ takes in 9 arguments while ‘niche_LR_cell’ has one
additional argument.</p>
<details><summary>
Arguments
</summary><ul>
<li>object: Niche-DE object</li>
<li>ligand cell: The cell type that expresses the ligand</li>
<li>receptor cell: The cell type that expresses the receptor</li>
<li>ligand_target_matrix: A matrix that measures the association between
ligands and their downstream target genes. The dimension should be
#target genes by #ligands</li>
<li>lr-mat: A matrix that matches ligands with their corresponding
receptors. This matrix should have two columns. The first will be
ligands and the second will be the corresponding receptors</li>
<li>K: The number of downstream target genes to use when calculating the
ligand activity score. Default value is 25.</li>
<li>M: The maximum number of ligands that can pass initial filtering.
Default value is 50.</li>
<li>alpha: The level at which to perform the Benjamini Hochberg
correction. Default value is 0.05.</li>
<li>truncation value: The value at which to truncate T statistics.</li>
</ul>
For ‘niche_LR_cell’, their is an additional parameter ‘alpha_2’ which
refers to the null quantile to compare observed epression to. That is
that we compare expression of the ligand and receptor to see if it is
greater than the ‘alpha_2’ percentile of gene expression over all genes
expressed in the ligand expressing or receptor expressing cell.The
Default value is 0.5 (50th percentile).
</details><p>We now perform ligand-receptor analysis to infer ligand-receptor
interactions between tumor cells (ligand expressing cell) and
fibroblasts (receptor expressing cell) on spot level data. The output
will be a list of ligands and their corresponding receptors. Note that
the ligand-target potential matrix refers to human genes.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"niche_net_ligand_target_matrix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ramilowski_ligand_receptor_list"</span><span class="op">)</span></span>
<span><span class="va">fibro_tumor_LR</span> <span class="op">=</span> <span class="fu"><a href="../reference/niche_LR_spot.html">niche_LR_spot</a></span><span class="op">(</span><span class="va">NDE_obj</span>,ligand_cell <span class="op">=</span> <span class="st">'tumor_epithelial'</span>,receptor_cell <span class="op">=</span> <span class="st">'stromal'</span>,</span>
<span>ligand_target_matrix <span class="op">=</span> <span class="va">niche_net_ligand_target_matrix</span>,</span>
<span>lr_mat <span class="op">=</span> <span class="va">ramilowski_ligand_receptor_list</span>,K <span class="op">=</span> <span class="fl">25</span>,M <span class="op">=</span> <span class="fl">50</span>,alpha <span class="op">=</span> <span class="fl">0.05</span>,truncation_value <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#preview output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fibro_tumor_LR</span><span class="op">)</span></span></code></pre>
<p>The output should resemble a 3 column table of ligands,their
corresponding receptors, and a list of the top 5 downstream niche-DE
genes that drive the ligand potential scoring.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
